byway/src/main/webapp/routepage.js

```diff
@@ -12,40 +12,54 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-
-// copies of recommendations and selected stops, stored as arrays for synchronous updating
-let recs = [];
-let stops = []; // TODO: Make set to avoid duplicates
-
+ 
+// copies of recommendations and selected stops, stored as sets for synchronous updating
+let recs = new Set();
+let stops = new Set();
+ 
+// location representations of stops array
+let waypoints = new Set();
```

Comment:
Try to avoid storing duplicate data. It's best to have a 'single source of truth' and just transform or extract relevant parts of that data as needed, rather than trying to maintain independent copies of the data with different version so f the same information. If there's extra information about the stops, just store it in the form of waypoints, and when you just want the 'stops' values, extract just those. If there's no extra information and `waypoints` is just a different representation of the same information, just save `stops` and compute waypoints as needed. For example, I notice that whenever you use `waypoints` you use `Array.from` anyway, so it's not like you're avoiding constructing a new array every time by doing this.

`Array.from(stops).map(stop => ({location: stop}))` would construct a pretty-much equivalent array without having to store duplicate data.

Reply:
Thanks for reminding me of this method, I had a hard time Googling how to do something like this. Will change.